org.eclipse.jdt.core/compiler/org/eclipse/jdt/internal/compiler/lookup/CompilationUnitScope.java

```diff
@@ -240,16 +241,42 @@ void checkAndSetImports() {
 	resolvedImports[0] = getDefaultImports()[0];
 	int index = 1;
 
-	nextImport : for (int i = 0; i < numberOfStatements; i++) {
+	Predicate<ImportReference> isStaticImport = i -> i.isStatic();
+	Predicate<ImportReference> isNotStaticImport = Predicate.not(isStaticImport);
+
+	// GitHub 269: resolve non-static imports first, so that cyclic static imports can be resolved correctly
+	index = resolveImports(numberOfStatements, resolvedImports, index, isNotStaticImport);
+
+	// non-static imports are resolved now, "store" them before continuing with static imports
+	ImportBinding[] temp = new ImportBinding[index];
+	System.arraycopy(resolvedImports, 0, temp, 0, index);
+	this.imports = temp;
+
+	// GitHub 269: non-static imports are resolved, now we can resolve static imports
+	index = resolveImports(numberOfStatements, resolvedImports, index, isStaticImport);
+
+	// shrink resolvedImports... only happens if an error was reported
+	if (resolvedImports.length > index) {
+		System.arraycopy(resolvedImports, 0, resolvedImports = new ImportBinding[index], 0, index);
+	}
+	this.imports = resolvedImports;
+}
+
+private int resolveImports(int numberOfStatements, ImportBinding[] resolvedImports, int index, Predicate<ImportReference> filter) {
+	for (int i = 0; i < numberOfStatements; i++) {
 		ImportReference importReference = this.referenceContext.imports[i];
+		if (!filter.test(importReference)) {
+			continue;
+		}
+
 		char[][] compoundName = importReference.tokens;
 
 		// skip duplicates or imports of the current package
 		for (int j = 0; j < index; j++) {
 			ImportBinding resolved = resolvedImports[j];
 			if (resolved.onDemand == ((importReference.bits & ASTNode.OnDemand) != 0) && resolved.isStatic() == importReference.isStatic()) {
 				if (CharOperation.equals(compoundName, resolvedImports[j].compoundName)) {
-					continue nextImport;
```

Comment:
Any reason this was not changed in your commit @iloveeclipse ?

Reply:
> Any reason this was not changed in your commit @iloveeclipse ?

Yes, this is the **only** reason why the label was needed - to jump out the inner loop and continue with next import.

All other places were simple "continue" cases.

Reply:
Please restore this

Reply:
> Yes, this is the **only** reason why the label was needed - to jump out the inner loop 

Ah sorry, I missed the outer loop.

Reply:
Me either...